<h3>Ćwiczenie 2. Brakujące wartości</h3>

<h5>Pobranie zbioru danych dotyczących rejsu Titanic ze strony: https://www.openml.org/d/40945</h5>

In [2]:
import pandas as pd
import numpy as np

<h5>Dokonanie wstępnej analizy pierwszych 20 wierszy</h5>

In [3]:
df = pd.read_csv('Titanic.csv')
df.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex     age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female      29      0      0     24160  211.3375       B5        S    2   
1     male  0.9167      1      2    113781    151.55  C22 C26        S   11   
2   female       2      1      2    113781    151.55  C22 C26        S    ?   
3     male      30      1      2    113781    151.55  C22 C26        S    ?   
4   female      25      1      2    113781    151.55  C22 C26        S    ?   
5     male      48      0      0     19952     26.55      E12        S    3   
6   female      63      1      0     13502   77.9583       D7        S   10   
7     male      39      0      0    112050         0      A36        S    ?   
8   female      53      2      0     11769   51.4792     C101        S    D   
9     male      71      0      0  PC 17609   49.5042        ?        C    ?   
10    male      47      1      0  PC 17757   227.525  C62 C64        C    ?   
11  female      18      1      0  PC 17757   227.525  C62 C64        C    4   
12  female      24      0      0  PC 17477      69.3      B35        C    9   
13  female      26      0      0     19877     78.85        ?        S    6   
14    male      80      0      0     27042        30      A23        S    B   
15    male       ?      0      0  PC 17318    25.925        ?        S    ?   
16    male      24      0      1  PC 17558  247.5208  B58 B60        C    ?   
17  female      50      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female      32      0      0     11813   76.2917      D15        C    8   
19    male      36      0      0     13050   75.2417       C6        C    A   

   body                        home.dest  
0     ?                     St Louis, MO  
1     ?  Montreal, PQ / Chesterville, ON  
2     ?  Montreal, PQ / Chesterville, ON  
3   135  Montreal, PQ / Chesterville, ON  
4     ?  Montreal, PQ / Chesterville, ON  
5     ?                     New York, NY  
6     ?                       Hudson, NY  
7     ?                      Belfast, NI  
8     ?              Bayside, Queens, NY  
9    22              Montevideo, Uruguay  
10  124                     New York, NY  
11    ?                     New York, NY  
12    ?                    Paris, France  
13    ?                                ?  
14    ?                    Hessle, Yorks  
15    ?                     New York, NY  
16    ?                     Montreal,

Dane składają się z 14 kolumn, między innymi takie jak: age, sex, name, survived. Wśród wartości są dane ilościowe oraz jakościowe. Widoczne braki w danych w postaci znaku '?'

<h5>Uzycie funkcji: pd.isnull().sum() oraz pd.isnull().mean w celu sprawdzenia brakujących danych.</h5>

Zbiór danych nalezy zmodyfikować poprzez zmianę znaku '?' reprezentującego braki w danych na wartość na.nan, aby zostały one wykryte przez funkcję .isnull()

In [4]:
df = df.iloc[:,:].replace('?', np.nan)
df.head(10)

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5       1         1                              Anderson, Mr. Harry    male   
6       1         1                Andrews, Miss. Kornelia Theodosia  female   
7       1         0                           Andrews, Mr. Thomas Jr    male   
8       1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9       1         0                          Artagaveytia, Mr. Ramon    male   

      age  sibsp  parch    ticket      fare    cabin embarked boat body  \
0      29      0      0     24160  211.3375       B5        S    2  NaN   
1  0.9167      1      2    113781    151.55  C22 C26        S   11  NaN   
2       2      1      2    113781    151.55  C22 C26        S  NaN  NaN   
3      30      1      2    113781    151.55  C22 C26        S  NaN  135   
4      25      1      2    113781    151.55  C22 C26        S  NaN  NaN   
5      48      0      0     19952     26.55      E12        S    3  NaN   
6      63      1      0     13502   77.9583       D7        S   10  NaN   
7      39      0      0    112050         0      A36        S  NaN  NaN   
8      53      2      0     11769   51.4792     C101        S    D  NaN   
9      71      0      0  PC 17609   49.5042      NaN        C  NaN   22   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON  
5                     New York, NY  
6                       Hudson, NY  
7                      Belfast, NI  
8              Bayside, Queens, NY  
9              Montevideo, Uruguay

Obliczneie ilości nan dla poszczególnych kolumn

In [5]:
nan_sum = df.isnull().sum()
nan_sum


pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

Oblicznie procentowej ilości nan w konkretnych kolumnach dla wszystkich danych w kolumnie

In [6]:
nan_sum/df.shape[0]*100 #df.shape[0] - liczba wierszy

pclass        0.000000
survived      0.000000
name          0.000000
sex           0.000000
age          20.091673
sibsp         0.000000
parch         0.000000
ticket        0.000000
fare          0.076394
cabin        77.463713
embarked      0.152788
boat         62.872422
body         90.756303
home.dest    43.086325
dtype: float64

Policzenie średniej ilości nan-ów w kazdej kolumnie - wynikiem będzie stosunek ilości nan-ów do wierszy w kolumnie z występującymi danymi

In [7]:
df.isnull().mean()

pclass       0.000000
survived     0.000000
name         0.000000
sex          0.000000
age          0.200917
sibsp        0.000000
parch        0.000000
ticket       0.000000
fare         0.000764
cabin        0.774637
embarked     0.001528
boat         0.628724
body         0.907563
home.dest    0.430863
dtype: float64

Obliczenie jaki procent w dataframe stanowią braki danych

In [8]:
nan_sum.sum()/df.size*100

21.035687002073555

<h5>Określenie typu braków danych: całkowicie przypadkowe (MCAR), przypadkowe (MAR), nieprzypadkowe (MNAR).</h5>

In [9]:
df['bodyNull'] = np.where(df['body'].isnull(), 1, 0)
df

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex     age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female      29      0      0   24160  211.3375       B5        S    2   
1       male  0.9167      1      2  113781    151.55  C22 C26        S   11   
2     female       2      1      2  113781    151.55  C22 C26        S  NaN   
3       male      30      1      2  113781    151.55  C22 C26        S  NaN   
4     female      25      1      2  113781    151.55  C22 C26        S  NaN   
...      ...     ...    ...    ...     ...       ...      ...      ...  ...   
1304  female    14.5      1      0    2665   14.4542      NaN        C  NaN   
1305  female     NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male    26.5      0      0    2656     7.225      NaN        C  NaN   
1307    male      27      0      0    2670     7.225      NaN        C  NaN   
1308    male      29      0      0  315082     7.875      NaN        S  NaN   

     body                        home.dest  bodyNull  
0     NaN                     St Louis, MO         1  
1     NaN  Montreal, PQ / Chesterville, ON         1  
2     NaN  Montreal, PQ / Chesterville, ON         1  
3     135  Montreal, PQ / Chesterville, ON         0  
4     NaN  Montreal, PQ / Chesterville, ON         1  
...   ...                              ...       ...  
1304  328                              NaN         0  
1305  NaN                              NaN         1  
1306  304                              NaN         0  
1307  NaN                              NaN         1  
1308  NaN                              NaN         1  

[1309 rows x 15 columns]

In [10]:
df.groupby(['survived'])['bodyNull'].mean()

survived
0    0.850433
1    1.000000
Name: bodyNull, dtype: float64

W szczegółowym opisie danych wyczytałem, ze body to Body Identification Number, co oznacza, ze ta kolumna powinna tyczyć się jedynie osób, które nie przezyly,
co obrazują powysze wyniki: 
- 1.000 - 100% brak danych (100% nan-ów) dla pasazerow, ktorzy przezyli (brak koniecznosci identyfikacji ciala)
- 0.850433 - 85% brak danych dla ludzi, ktorzy nie przezyli (ciała większości osób nie zostały znalezione)
>Powyzsze wyniki pozwalają wysunąć wnioski, ze mamy do czynienia z MNAR - braki w danych nie są przypadkowe.

Powyzsze dwie linie kodu zapisane w jednej:

In [11]:
df['body'].isna().groupby(df['survived']).mean()

survived
0    0.850433
1    1.000000
Name: body, dtype: float64

Obliczając procentowy brak danych dla całego zbioru danych, wyszedł on na poziomie 21%, co uniemozliwia nam usunięcie wszystkich wierszy z tego typu brakami.
Jak napisałem powyzej, w zbiorze mamy do czynienia z nieprzypadkowymi brakami danych, co równierz nalezy uwzględnić w analizie. W celu  usunięcia braków, nalzałoby się skupić na poszczególnych kolumnach i dla kazdej indywidualnie decydować czy mozemy usunać w niej wiersze z NaN

<h5>Sprawdzenie liczebności poszczególnych etykiet dla danych zmiennych
jakościowych</h5>

In [12]:
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

      age  sibsp  parch  ticket      fare    cabin embarked boat body  \
0      29      0      0   24160  211.3375       B5        S    2  NaN   
1  0.9167      1      2  113781    151.55  C22 C26        S   11  NaN   
2       2      1      2  113781    151.55  C22 C26        S  NaN  NaN   
3      30      1      2  113781    151.55  C22 C26        S  NaN  135   
4      25      1      2  113781    151.55  C22 C26        S  NaN  NaN   

                         home.dest  bodyNull  
0                     St Louis, MO         1  
1  Montreal, PQ / Chesterville, ON         1  
2  Montreal, PQ / Chesterville, ON         1  
3  Montreal, PQ / Chesterville, ON         0  
4  Montreal, PQ / Chesterville, ON         1

In [13]:
for i in df.columns:
    print('Liczba etykiet zmiennej ' + i + ': {}'.format(len(df[i].unique())))

Liczba etykiet zmiennej pclass: 3
Liczba etykiet zmiennej survived: 2
Liczba etykiet zmiennej name: 1307
Liczba etykiet zmiennej sex: 2
Liczba etykiet zmiennej age: 99
Liczba etykiet zmiennej sibsp: 7
Liczba etykiet zmiennej parch: 8
Liczba etykiet zmiennej ticket: 929
Liczba etykiet zmiennej fare: 282
Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej embarked: 4
Liczba etykiet zmiennej boat: 28
Liczba etykiet zmiennej body: 122
Liczba etykiet zmiennej home.dest: 370
Liczba etykiet zmiennej bodyNull: 2


<h5> Wyświetlenie z użyciem funkcji print liczby wszystkich pasażerów.</h5>

In [14]:
print('Liczba wszystkich pasazerow: ', len(df.name.unique()))

Liczba wszystkich pasazerow:  1307


<h5>Wykorzystanie podobny sposób jak w punkcie 1.</h5>

In [15]:
print('Liczba wszystkich pasazerow: {}'.format(len(df['name'].unique())))

Liczba wszystkich pasazerow: 1307


<h5>Komentarz wyników otrzymanych w punkcie 1 i 2 oraz podział zmiennych ze
względu na dużą i małą moc zbioru (kardynalność).</h5>

Na podstawie podpunktu 1 i 2 zauwazyłem, ze dane dziela sie na etykiety do wartości 30 oraz powyzej 99, dlatego tez według takiego kryterium podziele je na zbiory

In [16]:
card_1=[] #duza moc zbioru
card_2=[] #mala moc zbioru
for i in df.columns:
    np.where(len(df[i].unique()) >= 30, card_1.append(df[i]), card_2.append(df[i]))
# print(card_1)[1,:]

<h5>Sprawdzeniue ilości unikalnych etykiet dla zmiennej mówiącej o kabinie danego
pasażera. </h5>

In [17]:
cabin_label = df.cabin.unique()
print("Liczba unikalnych etykiet: " + str(len(cabin_label)))
print("Typ zmiennej cabin_label: " + str(type(cabin_label)))

Liczba unikalnych etykiet: 187
Typ zmiennej cabin_label: <class 'numpy.ndarray'>


Pandas DataFrame jest przeznaczone do przechowywania danych tabelarycznych, natomiast NumPy array słuzy do przechowywania danych numerycznych.

<h5>Redukcja cech dla zmiennej opisującej kabiny poprzez zastąpienie
obecnych etykiet w formacie LL11 do etykiet zawierających tylko
pierwszą literę.</h5>

In [18]:
df['CabinReduced'] = np.where(df['cabin'].isnull(), np.nan, df['cabin'].astype(str).str[0])
df[['cabin', 'CabinReduced']].head(20)


cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9       NaN          NaN
10  C62 C64            C
11  C62 C64            C
12      B35            B
13      NaN          NaN
14      A23            A
15      NaN          NaN
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

<h5>Wyświetlenie (jak w punkcie 1) liczby etykiet dla zmiennych z punktu wyzej.</h5>

In [19]:
print('Liczba etykiet zmiennej CabinReduced: {}'.format(len(df.CabinReduced.unique())))

Liczba etykiet zmiennej CabinReduced: 9


In [20]:
100 - len(df['CabinReduced'].unique())/len(df['cabin'].unique())*100


95.18716577540107

Zredukowano kardynalność zbioru zmiennej opisującej kabiny o 95.18716577540107%

Taka redukcja znacznie ułatwi i przyspieszy nam analizę zbioru danych, co prawda kosztem utraty szczegółowości danych, ale mimo wszystko przy zachowaniu kluczowej inforamcji - litery mówiącej o danej klasie. Mowa w tym przypadku o redukcji az o
95.18716577540107% w stosunku do zbioru pierwotnego. 

In [21]:
df.to_csv('Titanic_modified.csv', index=False) #index=False jeśli nie chchemy zapisywać indeksów do pliku